File: src/main/java/ludwig/workspace/UsageTracker.java
package ludwig.workspace;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.SetMultimap;
import ludwig.changes.InsertReference;
import ludwig.model.NamedNode;
import ludwig.model.ReferenceNode;

import java.util.Set;

public class UsageTracker {
    private final SetMultimap<NamedNode, ReferenceNode> refs = HashMultimap.create();

    public UsageTracker(Workspace workspace) {
        workspace.changeListeners().add(change -> {
            if (change instanceof InsertReference) {
                InsertReference ref = (InsertReference) change;
                refs.put(workspace.node(ref.ref()), workspace.node(ref.id()));
            }
        });
    }

    public Set<ReferenceNode> usages(NamedNode<?> node) {
        return refs.get(node);
    }

}

File: src/main/kotlin/ludwig/workspace/UsageTracker.kt
package ludwig.workspace

import com.google.common.collect.HashMultimap
import ludwig.changes.Value
import ludwig.model.Node
import ludwig.model.SymbolNode

class UsageTracker(workspace: Workspace) {
    private val refs = HashMultimap.create<Node, SymbolNode>()

    init {
        workspace.changeListeners().add(changeListener { change ->
            if (change is Value && workspace.node(change.nodeId) is SymbolNode) {
                refs.put(workspace.node(change.value), workspace.node(change.nodeId) as SymbolNode)
            }
        })
    }

    fun usages(node: Node): Set<SymbolNode> {
        return refs.get(node)
    }

}
